File: india-portwatch-terminal/src/data/news.ts
import type { NewsEvent } from "@/types/portwatch";

export const newsEvents: NewsEvent[] = [
  { id: "NW-0841-HRMZ", timestamp: "08:41", source: "Reuters", tag: "HORMUZ", entity: "HORMUZ", severity: "elevated", sentiment: -0.42, text: "Tanker rerouting continues near Strait of Hormuz; premiums up 4.2%.", affectedPorts: ["INMUN", "INIXY", "INBOM", "INNSA", "INCOK"], confidence: 0.91 },
  { id: "NW-0828-MAA", timestamp: "08:28", source: "Splash247", tag: "CHENNAI", entity: "CHENNAI", severity: "severe", sentiment: -0.28, text: "Chennai berth 4 dredging delay extends anchorage queue to 24h.", affectedPorts: ["INMAA", "INENR"], confidence: 0.87 },
  { id: "NW-0812-BAB", timestamp: "08:12", source: "gCaptain", tag: "RED SEA", entity: "RED SEA", severity: "watch", sentiment: -0.36, text: "Two additional carriers pause Red Sea transits through Bab-el-Mandeb.", affectedPorts: ["INMUN", "INNSA", "INCOK"], confidence: 0.84 },
  { id: "NW-0755-BAY", timestamp: "07:55", source: "IMD", tag: "BAY", entity: "BAY OF BENGAL", severity: "watch", sentiment: -0.22, text: "Low-pressure system intensifying over south-east Bay of Bengal.", affectedPorts: ["INMAA", "INENR", "INVTZ", "INPRT"], confidence: 0.89 },
  { id: "NW-0739-BOM", timestamp: "07:39", source: "Bloomberg", tag: "MUMBAI", entity: "MUMBAI", severity: "watch", sentiment: -0.14, text: "JNPT throughput slips 6% W/W amid labour rotation and monsoon swell.", affectedPorts: ["INBOM", "INNSA"], confidence: 0.83 },
  { id: "NW-0720-MUN", timestamp: "07:20", source: "PortNews", tag: "MUNDRA", entity: "MUNDRA", severity: "normal", sentiment: 0.08, text: "Mundra resumes normal ops after fog window at 05:40 UTC.", affectedPorts: ["INMUN"], confidence: 0.78 },
];

export const alertEvents = [
  { id: "AL-2041", portCode: "INMAA", severity: "severe", text: "Berth queue > 22h; recommend ETA shift +36h", ts: "08:44" },
  { id: "AL-2039", portCode: "INBOM", severity: "severe", text: "Congestion 0.83 - HSMM regime CONGESTED_HIGH", ts: "08:31" },
  { id: "AL-2036", portCode: "INNSA", severity: "severe", text: "Yard utilisation 94% - reroute cont. traffic", ts: "08:22" },
  { id: "AL-2033", portCode: "INHAL", severity: "watch", text: "River pilotage window narrowing (tide)", ts: "08:10" },
  { id: "AL-2029", portCode: "INPRT", severity: "watch", text: "SAR confidence 0.52 - cloud cover 88%", ts: "07:52" },
  { id: "AL-2025", portCode: "INVTZ", severity: "watch", text: "Cyclogenesis probability 34% (T+72h)", ts: "07:31" },
] as const;

export const entitySentiment = [
  { entity: "HORMUZ", mentions: 42, sentiment: -0.42 },
  { entity: "RED SEA", mentions: 31, sentiment: -0.36 },
  { entity: "CHENNAI", mentions: 24, sentiment: -0.28 },
  { entity: "MUMBAI", mentions: 19, sentiment: -0.14 },
  { entity: "MUNDRA", mentions: 12, sentiment: 0.08 },
  { entity: "SUEZ", mentions: 8, sentiment: -0.06 },
  { entity: "MALACCA", mentions: 6, sentiment: 0.02 },
];

File: backend/app/routes/news.py
from __future__ import annotations

from fastapi import APIRouter

router = APIRouter()


EVENTS = [
    {
        "id": "NEWS-001",
        "timestamp": "08:41",
        "source": "PortWatch AI",
        "tag": "CHENNAI",
        "entity": "CHENNAI",
        "severity": "elevated",
        "sentiment": -0.24,
        "text": "Monsoon-linked wind and rainfall signals may affect pilotage windows near Chennai.",
        "affectedPorts": ["INMAA"],
        "confidence": 0.82,
    },
    {
        "id": "NEWS-002",
        "timestamp": "08:31",
        "source": "PortWatch AI",
        "tag": "JNPT",
        "entity": "JNPT",
        "severity": "watch",
        "sentiment": -0.18,
        "text": "Container yard utilization remains elevated but manageable at JNPT.",
        "affectedPorts": ["INNSA"],
        "confidence": 0.78,
    },
    {
        "id": "NEWS-003",
        "timestamp": "08:20",
        "source": "PortWatch AI",
        "tag": "MUNDRA",
        "entity": "MUNDRA",
        "severity": "normal",
        "sentiment": 0.05,
        "text": "Demand pressure steady with moderate export flow at Mundra.",
        "affectedPorts": ["INMUN"],
        "confidence": 0.76,
    },
    {
        "id": "NEWS-004",
        "timestamp": "08:12",
        "source": "PortWatch AI",
        "tag": "HORMUZ",
        "entity": "HORMUZ",
        "severity": "elevated",
        "sentiment": -0.31,
        "text": "Tanker rerouting risk remains elevated near Strait of Hormuz.",
        "affectedPorts": ["INMUN", "INNSA"],
        "confidence": 0.80,
    },
]


ALERTS = [
    {
        "id": "AL-001",
        "portCode": "INMAA",
        "severity": "severe",
        "text": "Berth queue > 22h; recommend ETA shift +36h",
        "ts": "08:44",
    },
    {
        "id": "AL-002",
        "portCode": "INNSA",
        "severity": "severe",
        "text": "Container yard utilization elevated; monitor gate flow",
        "ts": "08:31",
    },
    {
        "id": "AL-003",
        "portCode": "INMUN",
        "severity": "watch",
        "text": "Weather window may affect pilotage schedule",
        "ts": "08:20",
    },
]


@router.get("/news")
def news_bundle() -> dict:
    return {
        "events": EVENTS,
        "alerts": ALERTS,
        "summary": {
            "totalEvents": len(EVENTS),
            "totalAlerts": len(ALERTS),
            "negativeEvents": sum(1 for e in EVENTS if e["sentiment"] < 0),
            "averageConfidence": round(
                sum(e["confidence"] for e in EVENTS) / len(EVENTS), 3
            ),
        },
    }


@router.get("/news/entity/{entity}")
def news_for_entity(entity: str) -> dict:
    entity_lower = entity.lower()

    matches = [
        e
        for e in EVENTS
        if entity_lower in e["entity"].lower()
        or entity_lower in e["text"].lower()
        or any(entity_lower in p.lower() for p in e["affectedPorts"])
    ]

    return {
        "events": matches,
        "alerts": [
            alert
            for alert in ALERTS
            if entity_lower in alert["portCode"].lower()
            or entity_lower in alert["text"].lower()
        ],
    }
